### ISSUE TITLE
FRQ: Horizontal Scrolling Option [enhancement]

### ISSUE THREAD

**Issue by njyo** [noah-nuebling/mac-mouse-fix#71]
Hi @noah-nuebling, nice little tool you've made here!

## Description
Could you add an additional option to set horizontal scroll speed independent from vertical scroll speed?

## Rationale
I'm using the [Apple Mighty Mouse](https://en.wikipedia.org/wiki/Apple_Mighty_Mouse) as I've not found a mouse to match its trackball in terms of 360º scrolling _(in theory)_.

Unfortunately, nowadays macOS (Big Sur) links the horizontal and vertical scrolling speed in a ratio, which makes either of the speeds unusable. I get to choose between: 😡

1. Usable horizontal scroll and supersonic vertical scroll or
2. Usable vertical scroll and grandpa with a walker horizontal scroll

I was hoping that you could extend Mac Mouse Fix with a separate horizontal scroll speed setting so that I can configure the two speeds separately and balance them to both be usable?

## Additional Info
Happy to sponsor you with a [(used) Apple Mighty Mouse](https://www.ebay.de/sch/i.html?_from=R40&_trksid=p2380057.m570.l1313&_nkw=mighty%20Mouse&_sacat=0), in case you'd need one, if that's a feature you'd be keen to add. 🥺 

PS: I'd also be happy with an (optional) menu bar icon showing that the tool is running.

**Comment by noah-nuebling** [comment]
Hi Wolfgang, thanks for your well-written feedback, and for offering to sponsor a Mighty Mouse!

I'd love to help you with your issue, but I probably won't get around to implementing something anytime soon as it’s pretty hard to do well. 

I've thought about implementing some alternative features though and I’d love to hear your thoughts on them:
1. A feature where you can hold a mouse button to scroll horizontally
2. A feature where you can drag websites and other content around by clicking and dragging a mouse button.

If you have any thoughts on how you think using these features with a 3rd party mouse would compare with your current workflow using the Mighty Mouse, feel free to let me know. Your input would be really valuable to inform my future decisions this.

---

I've thought about implementing your request and here's why I think it's hard to do well:

#### 1. I'm not sure as to how this would fit into the user interface. 

My thinking is that, you don't want to force people to configure the horizontal and vertical speeds separately, because for most people it's probably fine if they are the same. For most users, this would create (an admittedly pretty small amount of) extra mental load which we don't want.
Based on that I think it's probably a good idea to hide the option behind some sub-menu or behind pressing a keyboard modifier while starting the app or something similar. But I'm not sure where it would fit in exactly

#### 2. The technical aspect

While the feature you’d like wouldn’t be very hard to implement for _smooth scrolling_, implementing separate horizontal and vertical speeds for _non-smooth scrolling_, wouldn't be trivial to do. That‘s because non-smooth scrolling is almost entirely controlled by the system mouse driver. Mac Mouse Fox just takes its output and makes some minor adjustments like flipping the direction.

I can think of 3 options for making this work:
1. We'd reconfigure the system mouse driver to exhibit the desired behaviour. This is probably pretty hard in any case, and if the system mouse driver doesn't support separate horizontal speeds natively it would be completely impractical I recon.
2. Another way I see is to replicate the system mouse drivers’ scroll acceleration algorithm in Mac Mouse Fix to then be able do anything we want with it. Not easy either, unfortunately.
3. The last way I can think of is to reconfigure the system mouse driver on the fly. So we'd bascally turn the system mouse drivers sensitivity down, whenever a horizontal scroll event is detected by Mac Mouse Fix, and turn it back up when a vertical one is detected. This wouldn’t be super hard but it‘s kind of a hack and I'm not sure it would work well.

If you have any ideas on how to implement this well from a UI standpoint, please let me know! 
I’d also like to mention that I'm really unsure of my decisions and assessments when it comes to this design stuff, and I'm very much open to changing my opinion on any of this.

**Comment by noah-nuebling** [comment]
As far as the menu bar item goes - can you say why you would prefer having one? I personally like having as few menu bar items as possible, (for no practical reason at all - I just think it feels 'cleaner'), and I'm sort of blind to your perspective because of it. I'd love to change that though.

**Comment by njyo** [comment]
Hi @noah-nuebling,

Thanks for your elaborate response. Here a few more thoughts in this regard from my side:

### 1) Horizontal Scrolling Acceleration/Smoothness

Honestly, I'm not too concerned about the smoothness (or lack thereof) for the horizontal scroll, it's about the speed, which is just too low in Big Sur. I'm currently using [SmoothScroll Mac ](https://www.smoothscroll.net/mac/), but obviously, I'd prefer an open-source tool for such a system-level extension. Not sure how they solved it, but it looks like there's a way to do it. In the end, for me, it's really just a matter of mouse speed.

### 2) Preference Pane

As coincidence has it, I'm originally an Interaction Designer, so I should be able to help you along with the UI side of the preference pane. I did a quick mockup based on the Alfred preference pane (I'm more used to working on app flows, so I _didn't spend a lot_ of time optimising, but rather something OK enough).

![Mac Mouse Fix - Preference Pane - Mac Mouse Fix](https://user-images.githubusercontent.com/988408/113500375-017fb100-9550-11eb-8214-d9e9405d0ca2.png)

As you can see, you could even split things even more into tabs (like `General`, `Mouse Buttons` ansd `Scrolling`), but of course, then that might become confusing with the app-specific preferences. If so, I think you could so something more in line with the macOS keyboard shortcuts, where I can have a list of apps (with the web general settings on top) and their overriding settings.

![2021-04-04 at 14 17](https://user-images.githubusercontent.com/988408/113500521-34767480-9551-11eb-925f-d4aeb43aee30.png)


### 3) Menu Bar Item

While I personally prefer the menu bar clean and empty (and use Barkeeper for it), my main concern is knowing that things are running. Currently, I cannot notice that Mac Mouse Fix is actually running, as I didn't notice a difference when toggling the settings. Hence having an (optional) indication would be good. I think [App Trap](http://onnati.net/apptrap/) did this nicely, and so I emulated that in the above settings dialogue. :)

Hope that helps?

**Comment by noah-nuebling** [comment]
Hey Wolfgang, sorry for the late reply, and thanks so much for the detailed feedback and for taking the time to create a mockup! That's really cool of you to do. :)

The following are my thoughts on your points and the mockup mixed with some stream of conscience explanations of my thoughts surrounding the matter. I didn't think it would end up being THAT long, and I totally understand if you don't have time to read it. In that case I just want you to know that you've been very helpful, and that I appreciate it.

> ### Honestly, I'm not too concerned about the smoothness (or lack thereof) for the horizontal scroll...

I brought up the point that implementing a speed slider for non-smooth-scrolling would probably be much harder than doing it for smooth scrolling, because I thought that only having the horizontal speed slider available for smooth scrolling and not for non-smooth scrolling might be bad from a UX perspective. And in certain situations, I think that, if you can't do a feature "right", it's better not to do it at all. But upon thinking about this further, I don't think it's the end of the world to have no speed slider for non-smooth scrolling. In fact, the speed slider is disabled for non-smooth scrolling anyways in current versions of the app. 

### The Mockup

Here some question and thoughts about it:
- What was your thinking behind adding the green checkmark confirming that "Mac Mouse Fix is running"? (I now understand this after reading your point about the Menu Bar Item)
- What was your thinking behind Having 2 checkboxes, both "Enable Mac Mouse Fix" and "Launch Mac Mouse Fix at login"? What would be the role of each? (I now understand this after reading your point about the Menu Bar Item and taking a look at AppTrap)
- What is your thinking behind being able to choose whether Mac Mouse Fix shows up in the Dock?
- What is your reasoning behind turning the "Swap side button actions" as well as the "Smooth scrolling" and the "Scroll direction" checkboxes into popup buttons? (I think I understand this for "Scroll direction" now - it's so you can choose between "Natural" and "Standard" instead of having a "Invert direction" checkbox like you do now - good idea!)
- Mac Mouse Fix's options are (roughly) grouped by importance. Options that are very relevant and easy to understand, are front and center right when you open the app, while more niche or more complex options are hidden in a submenu (the "More..." menu) I hope that this simplifies the main interface and decreases mental load for most users who don't need the "niche" options. I only recently became fully aware that I'm implementing this design pattern, and I think part of it stems from the fact that Mac Mouse Fix used to be a preference pane and should be reconsidered. However if we want to apply this pattern, then, the horizontal scroll options most definitely don't belong on the main view. I say this under the assumption that well under 1% of users would want to use feature, and for them it would clutter things up. But maybe I'm overestimating the importance of this. At the end of the day the mockup you made is still quite simple and easy to read, so maybe the "cluttering effect" of it is not as important as I think.
- I really like that you used "Natural" for the scroll direction instead of "Inverted". That should be a little harder to implement vs the way it's done currently, but it's clearly much more intuitive.

> ### As you can see, you could even split things even more into tabs...

- That's a good idea. I'll definitely consider splitting things up into tabs more in the future. I just wonder if it's going to make usability worse compared to having a set of simple options presented to you on a single page. The main arguments I can think of:
	- Arguments for having more stuff on a single page
		- You have to click less to get to the options you want to get to
		- It's easier to recognise controls than to recall in which tab or submenu they are in
			- I remember being a little put off when using SteerMouse for the first time because it had so many tabs and submenus and for a lot of them I didn't understand what they even did. But that could also have to do with the general complexity and design of the program, and not with having tabs per-se.
		- I made some mockups yesterday and I ended up mocking up a page with tons of controls, which scrolls instead of being split up into tabs (see below). I've never seen that in a Mac App, though. So I think that maybe it's a bad design pattern, but I can't think of a concrete reason why it would be beyond that it's uncommon.
	- Arguments for splitting things up into tabs
		- The content of each tab is less and therefore easier to parse
		- You can easily navigate groups of options by clicking on tabs

__So my thinking on the UI layout boils down to this__

 There are two sensible ways in which all of the options of Mac Mouse Fix can be grouped: 
- By Importance (How often are they accessed, how many users find them relevant, etc.)
- By category (Like "Button settings", "Scroll settings", etc.)

And a big question is do we want to group our layout  _importance first_ (like it is now) or _category first_ (Like most other preference UIs).

I can think of the following ways to concretely implement this grouping:
- Proximity of UI elements on a single page
- Tabs 
- Submenus (Like the current "More..." menu)

> ### I think you could so something more in line with the macOS keyboard shortcuts, where I can have a list of apps...

That's an interesting idea. But if I understand correctly, that would mean that there would be a list of apps always present in the window of, no matter which tab you choose. Again, maybe this is a matter of me having suboptimal priorities (which I often do) but I think that that would clutter up the user interface too much. I think that app-specific settings are overall a pretty niche/power-user thing*, and therefore I'd want the option to be hidden in a submenu or at least not take up too much space on the main screen.

I want to elaborate on the stament from above "I think that app specific settings are overall a pretty niche/power-user thing*" 
→ Actually, The only reason the app-specific settings exist in their current form, is that the smooth-scrolling is not compatible with certain apps, as it's implemented. You can only use the app-specific settings to disable smooth-scrolling or keyboard modifiers as it stands. So I wouldn't even consider them a "power-user feature" in their current form, but a _crutch_ instead.

> ### 3) Menu Bar Item

That's a really good point, that I'm fully on board with. I agree that seeing which apps are currently running in the background from looking at the statusbar is very nice.

---

 You've brought many interesting ideas to the table and writing this up forced me think reconsider my thoughts and clarify them to you and to myself in the process. I think it's very easy to get stuck in your ways of thinking and run off faulty assumptions and priorities without every questioning or understanding them. You definitely helped me shake some of that.

So thanks a lot! :)

**Comment by noah-nuebling** [comment]
If you're interested here are the 

### Mockups I made yesterday:

for a new window to show up when you press the "More..." button

![About](https://user-images.githubusercontent.com/40808343/116003885-79497480-a600-11eb-9977-95fa19fcb10d.png)
![Settings](https://user-images.githubusercontent.com/40808343/116003889-7a7aa180-a600-11eb-894c-bd6e5ba5c4c6.png)

And here are my

### Thoughts on what is not ideal about these mockups, yet

**General**
* The split into "About" and "Settings"
	* Everything in the app is kind of settings. Alternative name ideas "Accessibility" or "Advanced"
* Symbol for "about" I don't like that much
* In normal apps, we'd access "About" and "Preferences" separately from the application menu. Do we have a reason to break this convention? (And go with a "More..." Button in the app which contains both "About" and "Preferences" instead)
* Do we need to add secondary labels or tooltips to clarify anything?
* Do we have too much whitespace making things look floaty?
	- The amount of whitespace looks ridiculous next to other apps
	- But if we reduce the vertical spacing it should be fine
		- I just reduced the spacing between the "Step size" and "Acceleration" sections from around 42 to 22, and that made the white space look reasonable

**About screen**
* Questionable: Symbol for Feature Request link
* Questionable: Symbol for Acknoledgements
* Questionable: Order of links
	- E.g. maybe it's bad psychologically, if file a bug report link is first
* Questionable: Placement of the Donation Button and the Acknoledgements
	- DonationButton should be humble and stay out of the way but still encouraging people to actually click on it
	- Acknoledgements should maybe have a special place and be displayed in-app not as a link (maybe)

**Settings screen**

* Questionable: We're lumping general settings like "Check for updates" in with advanced settings
* Questionable: This pattern of having a long scrolling list of toggles is something no-one else except me seems to design. Why is that? Would a series of tabs be better than the scrollview?
	* Maybe the scrollview is a little long? Especially when we have devices attached, and their device specific settings show up at the very bottom of the scrollview
* Should "App-specific settings" really be a popup button? - looks a little weird
* Keyboard modifier fields
	* Are the keyboard modifier fields too hard to parse due to too much text?
	* Why is the x button outside of the fields?
	* Is everyone (who cares about these features) going to understand the keyboard symbols? If not, Can we do anything to make stuff more clear?
	* How are users going to set keyboard mods for scrolling app-specifically??
* Is "tight" and "loose" really the best way to describe the smooth scrolling?
	- Ideas: Snappy, inertial, smooth, gliding,
* I think the horizontal step size settings make no sense
	* Do they actually solve Wolfgangs problem?
	* Why are they placed where they are?
* Settings that we could add maybe
	- Scrolling
		* Smooth scrolling on/off
		* Scrolling direction natural/standard
		* App-specific scroll settings

---

Again, I 100% understand if you don't have time to engage with this, but In case you have any feedback you want to share, I'd greatly appreciate that!

**Comment by njyo** [comment]
Hi @noah-nuebling, thanks a lot for your detailed response and the thought behind it!
I'll try to comment to the key parts.

### 1) Horizontal Speed

While I fundamentally agree with "better doing things right, or not at all", I also believe in "better have something OK, than nothing". In this case, it's really a matter of me currently having a significant problem with horizontal scrolling, where an OK solution will already be helpful. :)

### 2) Checkmark, enabling Mouse Fix and launching it on login

I think you know now, what each of them intend to do for me, but in case, here the recap:
- Checkmark indicates that Mouse Fix is running. This could be in the app or in the menu bar. But somewhere in the UI, I should be able to see it, so I can notice the difference of it on and off.
- Turning on/off Mouse Fix: Even if not launching at login, one should be able to turn the whole thing on or off. Unless the take is that as soon as the app is running (the window displayed), it is on. But that may be confusing.
- Launch at login is very useful and presumably will be enabled for anybody that finds the tool useful. Still, the option is a standard option and should be there.

### 3) Dock

Most utilities (e.g. Moom, Alfred, Tot, etc.) have some standard options:
- Menu bar only
- Dock only
- Both
- None (headless)

… personally I keep most menu bar only (so I can see that they are running), but I also like to have my menu bar clean. I don't know how many people use the Dock icon for any of these. So I personally don't find the Dock icon crucial and thus am OK without it.

### 4) Swap side buttons

Making it a drop-down makes the option clearer for me. Checkboxes are notorious for being not clear what ticking them has as a result. Also, drop-downs can make the UI look cleaner. That said, it's no biggie to have a checkbox there (I probably just copied it from the other options)

### 5) Grouping of options

I agree with you that important things should have more prominence and come first, though I would argue that grouping should be logical and overall structure should be top-down and thus order only affects peers. For example:

1. First order is about turning the app on and off
2. Then the next level is about the various functions (buttons, scrolling)
3. Within scrolling, vertical is much more important than horizontal, so should be first…

"Clutter is a failure of design, not an attribute of information." —E. Tufte ;)

In that sense, if grouped and ordered well, it won't seem cluttered, as users will be able to skim and glance.

### 6) Tabs

Agree that we should not do a death by tabs for a nifty little tool like this. And not mix tabs and sub-menus (like what you refer to with SteerMouse), but stick to one and keep it clean. I see no problem having 3-4 tabs (comparing with Apple's Trackpad preference pane):

- General Settings: General, Buttons, Scrolling, <Device Specific>
- Application Specific Settings: Own tab with the ability of add apps and configure changes there
- About

Yes, I agree with your argument for the tabs. :) 

### 7) App-Sepcific Settings

Thanks for the clarification. I'd think a separate tab would be fine, keeps them out of sight for most and most won't change them. Even if it's a crutch or power-user thing.

### 8) About and Settings

I would have the Settigns tab first and the About tab last and have the application ("Mac Mouse Fix") menu link to the specific tabs. Or, alternatively, you can even make them two separate windows, one for Settigns and one for About. Allows you some more flexibility for the About layout (e.g. single-column).

### 9) Mockups: About

This looks very nice. If I'd nitpick, I would consider the "Done" button to be centered in the bottom and consider merging the 6 links in the middle to 4 (merging help, website and GitHub) into a single website link… because I think most do fairly similar things… but that is honestly nitpicking. I think what you have there is 👌 

And the icon looks fine to me. So do the "questionable symbols". I'm sure there's more perfect ones, but they are good enough IMHO. Also, replacing symbols later is not too difficult I'd believe.

### 10) Mockups: Settings

Yes, this is a little airy. and long… mayber can be made a bit more compact. I'd also need to understand what the different things do to help grouping them better. Finding similarities will help with symmetry and rhyme, which will clean up things. For example: "always scroll precisely" and "horizontal step size matches vertical step size" could be used similarly to create "rhyme".

Overall a left-side anchor label with a right column of options might help keeping things tight. As well as using some options to hide/show sub-options (e.g. always scroll precisely "off" would probably hide certain detailed dials?).

The device-specific options could also be in a tab that's potentially disabled, or just displays "you don't have relevant devices attached" with al ink to supported special devices on the website.

App-Specific Settings I'd put in a tab with the take as outlined before. If not that, I'd need to think again.

I like "high inertia" and "low inertia" for the scrolling options, if that's doing what I think it does?

Yes, the horizontal step size should solve my problem. 🎉

---

Thanks a lot for taking your time and explaining your very thorough thoughts around this. I realise that I skipped some of the more nitty-gritty questions. I think once the overall framework becomes clear, than fitting all the options into it should be fine.
Maybe to support that, do you have somewhere a list of the features and settings and a description of what they do? That might help with it.

Thanks again!

**Comment by stevenla** [comment]
I'm partial to having a "A feature where you can drag websites and other content around by clicking and dragging a mouse button." option. I use a trackball (the Elecom Deft Pro), and being able to hold down one of my mouse buttons and spin the ball to pan around the page sounds really useful. I'm using Karabiner Elements to do this at the moment, since they have a `mouse_motion_to_scroll` option, but it's not smooth since it simulates a normal mouse wheel with discrete steps.

In design apps like Figma or Sketch, being able to pan around with 360° of freedom is helpful for quick navigation. I often need to pan at weird angles like 45° and usually just reach for my trackpad when I need to do that.
